### PR TITLE
vi script

### PR DESCRIPTION
**Pull request by yuanqing-wang**

### CODE REVIEW

**Comment by karalets** [comment]
Can you quickly tell me which model this is supposed to be exactly? 
How private/global are the latents?

**Comment by yuanqing-wang** [comment]
> Can you quickly tell me which model this is supposed to be exactly?
> How private/global are the latents?

This is the model with only one scalar as global variational parameters.

**Comment by karalets** [comment]
Where are the others?

**Comment by karalets** [comment]
Hello?

**Comment by yuanqing-wang** [comment]
@karalets , my biggest question is here:

https://github.com/yuanqing-wang/stag/blob/3f204fd9c5dcee31ac99f5b707b3594b1e5a8cf3/vi.py#L187

is the sum/mean correct here?

File: vi.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            p_a = self.a_prior
+
+            # variational posterior
+            q_a = torch.distributions.Normal(
+                loc=self.a_mu,
+                scale=self.a_log_sigma.exp()
+            )
+
+            # compute elbo
+            elbo = p_y_given_x_z.log_prob(y[mask]).mean()\
```
Comment (karalets) [inline vi.py]: Ok, let's consider the math for a second.

The variables Z, are they global or local? I.e. do we sample them separately for each datapoint, or just for each batch/forward-pass?

The general math if Z is global (and simplifying the multiple rounds  tensor thingie by writing capital Z) is:

logp(Y|X) >= log [ [Prod_i^{N} p(y_i| x_i,Z) ] * p(Z) / q(Z)] ,

Where N is the size of the DATASET, not the batch. So if we have minibatches we need to scale appropriately.
Note that Z is global and not inside the product over N.

IF Z is sampled separately per datapoint, we get:
logp(Y|X) >= log [ [Prod_i^{N} p(y_i| x_i,Z_i) * p(Z_i) / q(Z_i)] .

Note the difference that Z is now indexed by i and inside the product, which means it is a joint sum over the dataset.

What is currently implemented will not work in any way since it is neither of all these cases.

**Review by karalets — CHANGES_REQUESTED** [review]

File: vi.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        x : torch.Tensor
+            Input features. (n_nodes, in_features)
+
+        Returns
+        -------
+        dgl.DGLGraph
+            Graph with perturbed weights annotated.
+
+        """
+        g = g.local_var()
```
Comment (karalets) [inline vi.py]: please do sth like g_ = g.local_var() to ensure that we do not ovewrwrite anything outside, OR show me this is harmless with a little toy example.

File: vi.py (hunk 1 above)
Comment (yuanqing-wang) [reply vi.py]: - parameters: mu \in R(1); log_sigma \in R(1)
- sampling: sample one number Z for each feature, each edge, and each graph

**Comment by yuanqing-wang** [comment]
- name: STAG_VI(R)
- parameters: mu \in R(1); log_sigma \in R(1)
- sampling: sample two numbers for each feature, each edge, and each graph

File: vi.py (hunk 1 above)
Comment (yuanqing-wang) [reply vi.py]: - name of the model: STAG_VI(R)

Comment (yuanqing-wang) [reply vi.py]: in the loss function in L190, since it's broken down into different message passing rounds $Z$ would have the dimension: (n_edges, n_features)

Comment (yuanqing-wang) [reply vi.py]: E: number of edges, C: number of features, L: number of message-passing step
- STAG_VI(R^C): parameters mu \in R(C); log_sigma \in R(C); Z \in (L, E, C), independent in each message passing round;
- STAG_VI(R^E): neural network determines mu \in R(E); log_sigma \in R(E); Z \in (L, E, C), independent in each mp round;
- STAG_VI(R^(E \times C)): neural network determines mu \in R(E \times C); log_sigma \in R(E \times C); Z in Z \in (L, E, C)

File: vi.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        torch.Tensor : loss
+
+        """
+        # initialize losses
+        losses = []
+
+        for _ in range(n_samples):
+            # get input graph
+            # x.shape = (n_graphs, out_features)
+            g, x = self._forward(g, x)
```
Comment (karalets) [inline vi.py]: g_ !

Comment (karalets) [reply vi.py]: x_ !

File: vi.py (hunk 1 above)
Comment (yuanqing-wang) [reply vi.py]: replace `len(losses)` in L198 with `n_samples`

Comment (yuanqing-wang) [reply vi.py]: don't overwrite `g`

File: vi.py (hunk 3 above)
Comment (karalets) [reply vi.py]: if you want to simplify: pass x_ and g_ into the loss and do x, g = ....(x_, g_)
